graph/network/diffusion.go

```diff
@@ -0,0 +1,188 @@
+// Indexing into h is defined by the Laplacian Index field.
+//
+// Nodes without corresponding entries in h are given an initial heat of zero,
+// and entries in h without a corresponding node in the original graph are
+// not altered.
+func DiffuseToEquilibrium(h map[int64]float64, by Laplacian, tol float64, iters int) (eq map[int64]float64, ok bool) {
+	heat := make([]float64, len(by.Index))
+	for id, i := range by.Index {
+		heat[i] = h[id]
+	}
+	v := mat.NewVecDense(len(heat), heat)
+
+	last := make([]float64, len(by.Index))
+	for id, i := range by.Index {
+		heat[i] = h[id]
+	}
+	lastV := mat.NewVecDense(len(last), last)
+
+	var tmp mat.VecDense
+	for {
+		lastV, v = v, lastV
+		tmp.MulVec(by.Matrix, lastV)
+		v.SubVec(lastV, &tmp)
+		if normDiff(heat, last) < tol {
+			ok = true
+			break
+		}
+		iters--
+		if iters < 0 {
+			break
+		}
+	}
+
+	for i, n := range v.RawVector().Data {
+		h[by.Nodes[i].ID()] = n
+	}
+	return h, ok
+}
+
+// Laplacian is a graph Laplacian matrix.
+type Laplacian struct {
+	// Matrix holds the Laplacian matrix.
+	mat.Matrix
+
+	// Nodes holds the input graph nodes.
+	Nodes []graph.Node
+
+	// Index is a mapping from the graph
+	// node IDs to row and column indices.
+	Index map[int64]int
+}
+
+// NewLaplacian returns a Laplacian matrix for the simple undirected graph g.
+// The Laplacian is defined by D-A where D is a diagonal matrix holding the
+// degree of each node and A is the graph adjacency matrix of the input graph.
+func NewLaplacian(g graph.Undirected) Laplacian {
+	nodes := g.Nodes()
+	indexOf := make(map[int64]int, len(nodes))
+	for i, n := range nodes {
+		id := n.ID()
+		indexOf[id] = i
+	}
+
+	l := mat.NewSymDense(len(nodes), nil)
+	for j, u := range nodes {
+		to := g.From(u)
+		l.SetSym(j, j, float64(len(to)))
+		uid := u.ID()
+		for _, v := range to {
+			vid := v.ID()
+			if uid < vid {
+				l.SetSym(indexOf[vid], j, -1)
+			}
+		}
+	}
+
+	return Laplacian{Matrix: l, Nodes: nodes, Index: indexOf}
+}
+
+// NewSymNormLaplacian returns a symmetric normalized Laplacian matrix for the
+// simple undirected graph g.
+// The Laplacian is defined by I-D^(-1/2)AD^(-1/2) where D is a diagonal matrix holding the
+// degree of each node and A is the graph adjacency matrix of the input graph.
+func NewSymNormLaplacian(g graph.Undirected) Laplacian {
+	nodes := g.Nodes()
+	indexOf := make(map[int64]int, len(nodes))
+	for i, n := range nodes {
+		id := n.ID()
+		indexOf[id] = i
+	}
+
+	l := mat.NewSymDense(len(nodes), nil)
+	for j, u := range nodes {
+		to := g.From(u)
+		if len(to) == 0 {
+			continue
+		}
+		l.SetSym(j, j, 1)
+		uid := u.ID()
+		udeg := math.Sqrt(float64(len(to)))
+		for _, v := range to {
+			vid := v.ID()
+			if uid < vid {
+				l.SetSym(indexOf[vid], j, -1/(udeg*math.Sqrt(float64(len(g.From(v))))))
+			}
+		}
+	}
+
+	return Laplacian{Matrix: l, Nodes: nodes, Index: indexOf}
+}
+
+// NewRandomWalkLaplacian returns a random walk Laplacian matrix for the
```

Comment:
There is no description of `damp` in the documentation.

Reply:
I had intended to put it in, but forgot. I needed to work out how to add it into the equation. I will do that soon.

Reply:
Done.

Reply:
I don't see a description of `damp`?

Reply:
I wrote it and forgot to commit it.

Done